File: tools/python/scell_token_string_echo.py
#
# Copyright (c) 2017 Cossack Labs Limited
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys
from base64 import b64encode, b64decode
from pythemis import scell

_, COMMAND, KEY, MESSAGE, CONTEXT = range(5)

if len(sys.argv) not in (4, 5):
    print('Usage: <command: enc | dec > <key> <message> <context (optional)>')
    exit(1)

command = sys.argv[COMMAND]
key = sys.argv[KEY]
context = sys.argv[CONTEXT] if len(sys.argv) == 5 else ''

cell = scell.SCellTokenProtect(key.encode('utf-8'))
if command == 'enc':
    message = sys.argv[MESSAGE]
    encrypted, token = cell.encrypt(
        message.encode('utf-8'), context.encode('utf-8'))
    encoded_message = b64encode(encrypted).decode('ascii')
    encoded_token = b64encode(token).decode('ascii')
    print(','.join((encoded_message, encoded_token)))
elif command == 'dec':
    message, token = sys.argv[MESSAGE].split(',')
    decoded_message = b64decode(message.encode('utf-8'))
    decoded_token = b64decode(token.encode('utf-8'))
    decrypted = cell.decrypt(
        decoded_message, decoded_token, context.encode('utf-8'))
    print(decrypted.decode('utf-8'))
else:
    print('Wrong command, use "enc" or "dec"')
    exit(1)

File: tools/go/scell_token_string_echo.go
/*
 * Copyright (c) 2017 Cossack Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"encoding/base64"
	"fmt"
	"os"
	"strings"

	"github.com/cossacklabs/themis/gothemis/cell"
	"github.com/cossacklabs/themis/gothemis/keys"
)

func main() {
	args := os.Args

	if len(args) < 4 || len(args) > 5 {
		fmt.Printf("usage:\n\t%s {enc|dec} <key> <message[,token]> [context]\n", os.Args[0])
		os.Exit(1)
	}

	command := args[1]
	key := args[2]

	messageAndToken := strings.Split(args[3], ",")
	message := messageAndToken[0]
	var token string
	if len(messageAndToken) == 2 {
		token = messageAndToken[1]
	}

	var context string
	if len(args) == 5 {
		context = args[4]
	}

	sc, err := cell.TokenProtectWithKey(&keys.SymmetricKey{Value: []byte(key)})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Invalid key: \"%s\"", key)
		os.Exit(1)
	}

	switch command {
	case "enc":
		encrypted, token, err := sc.Encrypt([]byte(message), []byte(context))
		if nil != err {
			fmt.Fprintln(os.Stderr, "Error encrypting message")
			os.Exit(1)
		}
		fmt.Println(base64.StdEncoding.EncodeToString(encrypted) + "," + base64.StdEncoding.EncodeToString(token))

	case "dec":
		decodedMessage, err := base64.StdEncoding.DecodeString(message)
		if nil != err {
			fmt.Fprintln(os.Stderr, "Error decoding message")
			os.Exit(1)
		}
		decodedToken, err := base64.StdEncoding.DecodeString(token)
		if nil != err {
			fmt.Fprintln(os.Stderr, "Error decoding token")
			os.Exit(1)
		}
		decrypted, err := sc.Decrypt(decodedMessage, decodedToken, []byte(context))
		if nil != err {
			fmt.Fprintln(os.Stderr, "Error decrypting message")
			os.Exit(1)
		}
		fmt.Println(string(decrypted))

	default:
		fmt.Fprintln(os.Stderr, "Wrong command, use \"enc\" or \"dec\"")
		os.Exit(1)
	}
}
